src/Lecture3.hs

```diff
@@ -82,7 +81,39 @@ Tuesday
   would work for **any** enumeration type in Haskell (e.g. 'Bool',
   'Ordering') and not just 'Weekday'?
 -}
-next = error "TODO"
+instance Bounded Weekday where
+  minBound :: Weekday
+  minBound = Monday
+
+  maxBound :: Weekday
+  maxBound = Sunday
+
+-- In order to not depend on the declaration order I can not let the compiler
+-- auto-derive the instance but do it manually
+instance Enum Weekday where
+  toEnum :: Int -> Weekday
+  toEnum = \case
+    1 -> Monday
+    2 -> Tuesday
+    3 -> Wednesday
+    4 -> Thursday
+    5 -> Friday
+    6 -> Saturday
+    7 -> Sunday
+    _ -> error "enum out of bound"
```

Comment:
Deriving the `Enum` here actually is the way to go 👍🏻 

GHC uses consequent numbers for consequence constructors. If the numbers are not the part of the user interface and just an internal implementation detail, you can safely derive them.

Generally, all instances derived by GHC out-of-the-box are better be derived. Custom implementations of those instances are usually more error-prone and often provide more unexpected behaviour than the derived one. If you want a custom implementation, you can create a separate independent function 🙂